presto-native-execution/presto_cpp/main/common/PrometheusStatsReporter.h

```diff
@@ -0,0 +1,119 @@
+ public:
+  PrometheusStatsReporter(
+      const std::string cluster = "",
+      const std::string worker = "") {
+    if (cluster.empty()) {
+      auto nodeConfig = facebook::presto::NodeConfig::instance();
+      cluster_ = nodeConfig->nodeEnvironment();
+    } else {
+      cluster_ = cluster;
+    }
+    char* hostName = std::getenv("HOSTNAME");
+    workerPod_ = !hostName ? worker : hostName;
+  }
+
+  /// Register a stat of the given stat type.
+  /// @param key The key to identify the stat.
+  /// @param statType How the stat is aggregated.
+  void registerMetricExportType(
+      const char* key,
+      facebook::velox::StatType statType) const override;
+
+  void registerMetricExportType(
+      folly::StringPiece key,
+      facebook::velox::StatType statType) const override;
+
+  void registerHistogramMetricExportType(
+      const char* /*key*/,
+      int64_t /* bucketWidth */,
+      int64_t /* min */,
+      int64_t /* max */,
+      const std::vector<int32_t>& /* pcts */) const override {}
+
+  void registerHistogramMetricExportType(
+      folly::StringPiece /* key */,
+      int64_t /* bucketWidth */,
+      int64_t /* min */,
+      int64_t /* max */,
+      const std::vector<int32_t>& /* pcts */) const override {}
+
+  void addMetricValue(const std::string& key, size_t value = 1) const override;
+
+  void addMetricValue(const char* key, size_t value = 1) const override;
+
+  void addMetricValue(folly::StringPiece key, size_t value = 1) const override;
+
+  virtual void addHistogramMetricValue(const std::string& key, size_t value)
+      const override {}
+
+  virtual void addHistogramMetricValue(const char* key, size_t value)
+      const override {}
+
+  virtual void addHistogramMetricValue(folly::StringPiece key, size_t value)
+      const override {}
+
+  const facebook::velox::StatType getRegisteredStatType(
+      const std::string& metricName) {
+    std::lock_guard<std::mutex> lock(mutex_);
+    return registeredStats_[metricName];
+  }
+
+  /*
+   * Serializes the metrics collected so far in the format suitable for
+   * back filling Prometheus server.
+   *
+   * Given a metric name and a set of labels, time series are frequently
+   * identified using this notation:
+   *
+   *       <metric name>{<label name>=<label value>, ...}
+   *
+   * For example, a time series with the metric name num_tasks_aborted
+   * and the labels cluster="<cluster_id>" and worker="worker-id"
+   * could be written like this:
+   *  # HELP num_tasks_aborted
+   *  # TYPE num_tasks_aborted gauge*
+   *  num_tasks_aborted{cluster="<cluster_id>", worker="worker-id"} value
+   * timestamp
+   *
+   * Above info is from:
+   * https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels
+   */
+  const std::string getMetricsForPrometheus();
+
+ private:
+  /// Mapping of registered stats key to StatType.
+  mutable std::unordered_map<std::string, facebook::velox::StatType>
```

Comment:
Have you considered using https://github.com/jupp0r/prometheus-cpp or some opensource implementation for stats collection? If not any specific reason?

Reply:
On the other hand, it supports all metrics, is production tested and widely used (though like an unofficial library)

Reply:
I have created a draft PR - https://github.com/prestodb/presto/pull/21753, we believe that using the library is the fastest way to production since its production tested. Not sure how to integrate

Reply:
Using the library seems better if it is production-ready and widely used.
This can be an optional dependency that we can turn on/off at compile time.
I can help with the dependency management.

Reply:
Exactly, yes we can control that. Do you guys have a timeline for this work? I am already working on this on our end and we can collaborate

Reply:
@jaystarshot @majetideepak Lm take a look and try it out and integrate with this PR itself.
@jaystarshot would you mind pausing your work for a bit. I will reach out if I need help. Thanks.

Reply:
`is production tested and widely used`

@jaystarshot can you share some details on this?

Reply:
@karteekmurthys  Unfortunately this is a P0 from our side

Reply:
https://www.youtube.com/watch?v=EAS-3g9rqzo&ab_channel=CppUsergroupDresden
Seems like it is an opensource of googles bormon

Reply:
@karteekmurthys I believe Jay already has a PR in progress here https://github.com/prestodb/presto/pull/21753
We should base our work on that.
You can help with the dependency management in that PR.

Reply:
@jaystarshot I took a shot at using prometheus-cpp in this PR itself. I have added a compile time settings to include prometheus-directory. Added some unit tests. I will further modify the circle ci configs to run unit tests for this change. Feel free to use this PR for your testing. I will be opening an RFC soon to get this merged with Presto.

Reply:
Have you tried this in local? I don't think this will work since there is no exposer defined. My PR https://github.com/prestodb/presto/pull/21753 shows how to do that. Can you share the RFC with me too? I can also input from my end thanks

Reply:
We don't need an Exposer. We already expose an REST API, check PrestoServer.cpp changes in this PR.
You can configure your Prometheus server to scrape the enpoint: `/v1/info/health/metrics` exposed by the worker.
Or you can configure whatever intermediary process to scrape from worker endpoint.
Missed adding the test results in previous comment, here it is: [prometheus-format-metrics.txt](https://github.com/prestodb/presto/files/14117698/prometheus-format-metrics.txt)

Reply:
I see thanks! I think having an exposer is better since it separates the http server and makes it configurable based on the number of threads etc. In future we can also move this easily to a sidecar. We can sync offline regarding the RFC too if you have a draft prepared already